File: leetcode/medium/MergeTripletsToFormTarget.py
from typing import List

class Solution:
    """
    Determines if the target triplet can be formed by merging a subset of the
    given triplets.

    A merge operation takes two triplets (a, b, c) and (d, e, f) and produces
    (max(a, d), max(b, e), max(c, f)).

    The target can only be achieved if:
    1. Every triplet used in the merge has components less than or equal to the
       corresponding target components.
    2. Among the valid triplets, at least one achieves the target's first component,
       at least one achieves the second, and at least one achieves the third.
    """
    def mergeTriplets(self, triplets: List[List[int]], target: List[int]) -> bool:
        """
        Checks if the target triplet can be formed by merging valid triplets.

        Args:
            triplets: A list of triplets, where each triplet is a list/tuple of three integers.
            target: The target triplet [T1, T2, T3] to achieve.

        Returns:
            True if the target can be formed, False otherwise.
        """
        # Unpack the target for easy comparison
        T1, T2, T3 = target
        
        # Boolean flags to track if each component of the target has been matched
        # by at least one valid triplet.
        T1_achieved, T2_achieved, T3_achieved = False, False, False

        for f, s, t in triplets:
            # First filter condition: A triplet is 'valid' for merging only if
            # all its components are less than or equal to the target's components.
            # If any component exceeds the target, this triplet can *never* be
            # used, as the merge operation only increases values.
            if f <= T1 and s <= T2 and t <= T3:
                # This triplet is valid. Now check if it helps us achieve the target.
                
                # Check if this triplet *matches* the target's first component.
                if f == T1:
                    T1_achieved = True
                
                # Check if this triplet *matches* the target's second component.
                if s == T2:
                    T2_achieved = True
                
                # Check if this triplet *matches* the target's third component.
                if t == T3:
                    T3_achieved = True
            
            # Optimization: If all three target components have been matched,
            # we can stop early and return True immediately.
            if T1_achieved and T2_achieved and T3_achieved:
                return True

        # After checking all triplets, return True only if all three target components
        # were successfully matched by at least one valid triplet.
        return T1_achieved and T2_achieved and T3_achieved

File: leetcode/medium/MergeTripletsToFormTarget.java
class Solution {
    /**
     * Determines if a subset of 'triplets' can be merged (component-wise maximum) 
     * to form the 'target' triplet.
     * * The core logic is:
     * 1. A triplet A=(a1, a2, a3) can only be used if all its components are 
     * less than or equal to the corresponding target components (a1 <= t1, a2 <= t2, a3 <= t3).
     * If any component of A is greater than the target (e.g., a1 > t1), using A 
     * will make the merged result's first component > t1, making the target unachievable.
     * 2. To achieve the target T=(t1, t2, t3), we must find at least one **valid** * triplet that achieves t1 (i.e., [t1, a2, a3] where a2<=t2, a3<=t3), 
     * at least one valid triplet that achieves t2, and at least one valid 
     * triplet that achieves t3.
     * * @param triplets An array of [a, b, c] triplets.
     * @param target The target [t1, t2, t3] triplet.
     * @return true if a subset of triplets can be merged to form the target, false otherwise.
     */
    public boolean mergeTriplets(int[][] triplets, int[] target) {
        // Flags to track whether we have found a valid triplet that achieves each component
        boolean target1Achieved = false;
        boolean target2Achieved = false;
        boolean target3Achieved = false;

        final int t1 = target[0];
        final int t2 = target[1];
        final int t3 = target[2];

        for (int[] currentTriplet : triplets) {
            final int c1 = currentTriplet[0];
            final int c2 = currentTriplet[1];
            final int c3 = currentTriplet[2];

            // 1. Check for the Veto Condition: 
            // The triplet is only useful if all its components are <= the target's.
            if (c1 <= t1 && c2 <= t2 && c3 <= t3) {
                
                // 2. Track Achievement:
                // If a valid triplet matches a target component, set the corresponding flag.
                if (c1 == t1) {
                    target1Achieved = true;
                }

                if (c2 == t2) {
                    target2Achieved = true;
                }

                if (c3 == t3) {
                    target3Achieved = true;
                }

                // Optimization: If all three components are achieved, we can stop early.
                if (target1Achieved && target2Achieved && target3Achieved) {
                    return true;
                }
            }
        }
        
        // After checking all triplets, return true only if all components were achieved
        return target1Achieved && target2Achieved && target3Achieved;
    }
}